service/migration_manager.cc

```diff
@@ -132,6 +133,9 @@ void migration_manager::init_messaging_service()
                     "migration request handler: group0 snapshot transfer requested, but canonical mutations not supported");
             }
             cm.emplace_back(co_await db::system_keyspace::get_group0_history(proxy));
+            for (auto&& m : co_await replica::read_tablet_mutations(proxy)) {
+                cm.emplace_back(std::move(m));
+            }
```

Comment:
Why move it as part of schema and not topology mutations?

Reply:
When creating a table or dropping a table, I want tablet mutations and schema mutations to be applied atomically as part of one command. This way schema changes can be a single command.

Reply:
In other cases, like topology changes, tablet mutations can be part of topology mutations.

Reply:
Actually, we should prohibit pulling schema outside of raft if we care about local state machine consistency. It's not enough to pull tablets, we also need topology, on which those tablets depend, to properly choose replicas. For example, which ones are dc local.

If we're talking tablet mutations in the migration request, I can actually move them to the topology pull. It's just that it wasn't yet available at the time this was implemented.

Reply:
> Actually, we should prohibit pulling schema outside of raft if we care about local state machine consistency. It's not enough to pull tablets, we also need topology, on which those tablets depend, to properly choose replicas. For example, which ones are dc local.

Yes. We should at least do `raft_read_barrier`, recheck if schema is available and then pull. I think we still need to pull since the schema we want to pull may be the old one and not the latest one, no?

Reply:
But  should change the order, and sync raft first.

Reply:
What is the former? The one in maybe_sync? So what should it do?

```
co_await read_barrier()
return get_schema_definition()
```
?

Reply:
> What is the former? The one in maybe_sync? So what should it do?
> 
> ```
> co_await read_barrier()
> return get_schema_definition()
> ```
> 
> ?

Yes

Reply:
But we should also avoid read_barrier() if we already know the schema version.

Reply:
So what I wrote here https://github.com/scylladb/scylladb/pull/13387#discussion_r1175130799:

```
if (schema is not available) {
  co_await read_barrier()
}

if (schema is not available) {
  return get_schema_definition()
}
```

Reply:
> So what I wrote here [#13387 (comment)](https://github.com/scylladb/scylladb/pull/13387#discussion_r1175130799):
> 

get_schema_definition() does the check, so:

```
 if (auto s = local_schema_registry().get_or_null(v)) {
     return s; 
 }
 co_await read_barrier();
 return get_schema_definition(v, ...);
```

Reply:
And maybe_sync() becomes mark_synced(), since we did sync via raft.

Reply:
We actually can drop maybe_sync() altogether, since everything will be synced before being registered.

Reply:
> Why move it as part of schema and not topology mutations?

Pulling it as part of topology mutations is more tricky because we need to apply and reflect topology before the schema. So I propose we do it incrementally.